pwiz/data/vendor_readers/Waters/SpectrumList_Waters.cpp

```diff
@@ -722,6 +749,46 @@ PWIZ_API_DECL void SpectrumList_Waters::createIndex()
     size_ = index_.size();
 }
 
+PWIZ_API_DECL void SpectrumList_Waters::getDDAScan(unsigned int index, vector<float>& masses, vector<float>& intensities, IndexEntry& ie) const
+{
+    using namespace boost::spirit::karma;
+    
+    float setMass, precursorMass, retentionTime;
+    int function, startScan, endScan;
+    bool isMS1;
+    rawdata_->GetDDAScan(index, retentionTime, function, startScan, endScan, isMS1, setMass, precursorMass, masses, intensities);
+
+    ie.function = function;
+    ie.process = 0;
+    ie.block = -1; // The SDK DDA processor doesn't yet support ion mobility data
+    ie.scan = startScan; // While it might combine multiple scans, use the first for getting the metadata
+    ie.index = index;
+    ie.setMass = setMass;
+    ie.precursorMass = precursorMass;
+
+    ie.id = "";
+    std::back_insert_iterator<std::string> sink(ie.id);
+    if (startScan == endScan)
+    {
+        generate(sink,
+                "function=" << int_ << " process=" << int_ << " scan=" << int_,
+                ie.function + 1, ie.process, ie.scan + 1);
+    } 
+    else
+    {
+        generate(sink,
+                "function=" << int_ << " process=" << int_ << " scan=" << int_ << " merged=" << int_ << "-" << int_,
+                ie.function + 1, ie.process, ie.scan + 1, startScan + 1, endScan + 1);
+    }
+    idToIndexMap_[ie.id] = ie.index; 
+}
+
+PWIZ_API_DECL void SpectrumList_Waters::createDDAIndex()
+{
+    size_ = rawdata_->GetDDAScanCount();

```

Comment:
The index needs to be created up front. Accessing a single spectrum must mean that all spectra are indexed, which means their ids are generated and SpectrumList_Waters::spectrumIdentity() returns the IndexEntry instantly without having to allocate anything. If this is a costly operation, it can be progress monitored with iterationListenerRegistry. I don't think there's an example right now of using that in a vendor SpectrumList, but we do use it for a couple SpectrumListWrappers.

Reply:
Yes, I was also considering implementing it as a filter. However it seemed a bit complicated by the fact that the LockMassRefiner has a concrete dependency on SpectrumList_Waters, so all the Waters refiners would need to be applied in the right order, and probably aware of each other, so that seemed messier.

Reply:
Yes I considered that messiness but doesn't DDAProcessor replace LockMassRefiner (for DDA data, at least)?

Reply:
> Look to be a couple of issues with the build/checks. I think both are a result of the MassLynxSDK being built a different way to previous releases, presumably with different versions of the C++ runtime.
> 
> * We no longer have an x86 build. Is there a way to disable Waters raw support for the x86 build? Out of interest, is anyone still using x86?
> * There are some tiny changes in floating point numbers in the expected test results. Would you be happy with updating the expected results, provided we confirm all the changes are very small (they look to be in the 6th significant figure or thereabouts)?

Yes we can disable Waters for x86. Or possibly stick with the older SDK. Probably best to just disable though. I think the only x86 users these days would be for extremely old instrument computers. The AB Sciex 4700/4800 TOF/TOF has this weird API I access through their Data Explorer program and it's only available as x86. But yeah, extremely old.

Reply:
Yes, on further thought that does seem pretty complicated because normally all the special indexing logic would be kept in the filter, but in this case it would require close integration with the Waters API which we normally don't do in filters.